collector/zfs_linux.go

```diff
@@ -0,0 +1,83 @@
+package collector
+
+// +build linux
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/common/log"
+)
+
+const zfsArcstatsProcpath = "spl/kstat/zfs/arcstats"
+
+func zfsInitialize() error {
+	path := procFilePath(zfsArcstatsProcpath)
+	fd, err := os.Open(path)
+	if err != nil {
+		log.Errorf("Cannot open ZFS arcstats logpath: %s", path)
+		return err
+	}
+	return fd.Close()
+}
+
+func (p *zfsMetricProvider) PrepareUpdate() (err error) {
+
+	err = p.prepareUpdateArcstats(zfsArcstatsProcpath)
+	if err != nil {
+		return
+	}
+	return nil
+}
+
+func (p *zfsMetricProvider) handleFetchedMetricCacheMiss(s zfsSysctl) (value zfsMetricValue, err error) {
+
+	// all values are fetched in PrepareUpdate()
+	return zfsErrorValue, fmt.Errorf("sysctl '%s' found")
+}
+
+func (p *zfsMetricProvider) prepareUpdateArcstats(zfsArcstatsProcpath string) (err error) {
+
+	file, err := os.Open(procFilePath(zfsArcstatsProcpath))
+	if err != nil {
+		log.Errorf("Cannot open ZFS arcstats procfs file for reading. " +
```

Comment:
Should we decide per `Update()` call whether ZFS statistics are availabe or only on launch?

Pro only-on-launch: Less overhead on systems without ZFS.
Contra only-on-launch: If the ZFS module is loaded after node_exporter has started, node_exporter will require a restart.

The overhead of checking whether ZFS is available:

* Linux: Check if `/proc/spl/kstat/zfs/...` can be opened
* FreeBSD: Not implemented yet. Probably check for existence of ZFS sysctls.

All in all relatively low overhead, but this should be the *product owner's* decision.